### DISCUSSION TITLE
Relicensing Clarification

### DISCUSSION THREAD

**Discussion by theduke** [minio/minio#12158] [Q&A] [answered]
Hi there. 

In general I'm supportive of copyleft licenses and the AGPL for projects that are likely used in a context where the affero clauses become relevant.

However there is a question regarding the legality of this concrete change.

I'm sure you had legal counsel on this, but in my understanding, you can relicense contributions owned by the company as AGPL and subsume past external contributions. But without a CLA all external contributions remain licensed under Apache 2. 

Which means that the Apache 2 license needs to be retained to show that parts of the code base remain under Apache 2, and you can not just remove the Apache header from files where third parties have contributed without their permission . Which appears to have happened here.

Can you share your legal understanding on this issue?

**Comment by harshavardhana** [comment, marked as answer]
> Which means that the Apache 2 license needs to be retained to show that parts of the code base remain under Apache 2, and you can not just remove the Apache header from files where third parties have contributed without their permission . Which appears to have happened here.

@theduke all copyright headers in the code from external parties have been maintained that we have forked, whoever contributed were asked to add copyright headers pointing to MinIO for their contribution as part of the source header. 

We have always carried the NOTICE header in the project and there are no external contributions with external copyright by external organizations. Wherever applicable the headers are maintained such as the following forked/modified projects - all modifications are in AGPLv3.0 

https://github.com/minio/minio/blob/master/pkg/argon2/argon2.go#L38
https://github.com/minio/minio/blob/master/pkg/rpc/server.go#L6
https://github.com/minio/minio/blob/master/pkg/handlers/proxy.go#L18
https://github.com/minio/minio/tree/master/pkg/csvparser

The rest of them were always MinIO, so they have been simply changed to AGPL as per our requirement going forward.

**Reply by rjc** [reply]
This has been marked as _Answered_ but, in fact, does not address the question regarding external contributors. You've mentioned:

> We have always carried the NOTICE header in the project and there are no external contributions with external copyright by external organizations.

You're missing the point - copyright notice is usually added if a signifficant part of a file has been touched. As I've mentioned above, everyone who ever contributed a non-trivial change to your project would have to agree to the license change. Have you contacted all non-trivial contributors? You list [**12** people](https://github.com/orgs/minio/people), yet you have [many more non-trivial ones](https://github.com/minio/minio/graphs/contributors) not listed there. Are **all** of them liked to MinIO, or were they at the time? If not, short of rewriting all non-trivial 3rd-party (i.e. non-MinIO) contributions, would be to ask the contributors whether they're OK to relicense _their_ contribution.

**Comment by rjc** [comment]
@harshavardhana I don't think that's what @theduke meant. Every single non-trivial contribution up to the time you have changed the license, is covered by the old license. You can not re-license other people's contributions without them agreeing to it. Otherwise, you might be violating their copyright - https://opensource.stackexchange.com/questions/33/how-can-a-project-be-relicensed

**Comment by theduke** [comment]
Exactly, @rjc. I was not talking about vendored projects.

There are currently 295 contributors for this repository.

That means you would have had to either get written permission from each of those that are/were not employed by Minio, or completely remove their contributions.

It's certainly possible that you have done this, but it seems like a monumental task for an older code base.

Apache 2 is GPL compatible, so sublicensing to AGPL and subsuming the Apache parts should be OK, but in my understanding that doesn't mean you can just *RE*license all past contributions and remove Apache2.

**Reply by GarimaKapoor** [reply]
Thank you @theduke for raising your concerns. While we are excited to move to a copyleft license moving forward, we also acknowledge there are questions from the community on the contributions from the external contributors. We are going through the audit of trivial and nontrivial contributions and will make the process transparent to the community here. Please bear with us as we want to be very thorough with the review.

**Comment by marcan** [comment]
I'm also curious how MinIO plans to continue to offer commercial licensing. As they accept external contributions to MinIO, they cannot possibly offer the product under a different license any more after accepting external commits in the future, as they are required to comply with the AGPLv3 when distributing any version of MinIO that includes those external contributions.
